File: LeetCode/Hao/23MergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    public class PriorityComparator implements Comparator<ListNode>
    {
        public int compare(ListNode e1, ListNode e2)
        {
            return e1.val - e2.val;
        }
    }
    public ListNode mergeKLists(List<ListNode> lists) {
       
        if (lists == null)
        {
            return null;     
        }
        if (lists.size() == 0)
        {
            return null;
        }
        
        ListNode head = new ListNode(0);
        
        ListNode p = head;
        
        PriorityQueue<ListNode> q = new PriorityQueue<ListNode>(lists.size(), new PriorityComparator());
        
        
        for (ListNode list : lists)
        {
            if (list != null)
            {
                q.add(list);   
            }
        }
        
        
        while (q.size() > 0)
        {
            ListNode tmp = q.poll();
            p.next = tmp;
            
            if (tmp.next != null)
            {
                q.add(tmp.next);
            }
            
            
            p = p.next;
            
        }
        
        
        return head.next;
    }
}

File: LeetCode/Liang/23MergeKSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
/**
 * Try to use priority_queue to do it. And make it 
 **/
class Solution {
public:
    // Merge sort will not fit this. The time complexity is:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        return mergeKListsHelper(lists, 0, (int)lists.size() - 1);
    }
private:
    ListNode *mergeKListsHelper(vector<ListNode*> &lists, int begin, int end) {
        if (begin > end) {
            return nullptr;
        }
        
        if (begin == end) {
            return lists.at(begin);
        }
        
        // It seems that I forget this kind of process.
        int mid = begin + (end - begin) / 2;
        ListNode *left = mergeKListsHelper(lists, begin, mid);
        ListNode *right = mergeKListsHelper(lists, mid + 1, end);
        return mergeTwoSortedLists(left, right);
    }
    
    ListNode* mergeTwoSortedLists(ListNode *first, ListNode *second) {
        ListNode *dummy = new ListNode(0);
        ListNode *tail = dummy;
        while (first && second) {
            ListNode *tmp = nullptr;
            if (first->val < second->val) {
                tmp = first;
                first = first->next;
            } else {
                tmp = second;
                second = second->next;
            }
            
            tail->next = tmp;
            tail = tail->next;
        }
        tail->next = first ? first : second;
        first = dummy->next;
        delete dummy;
        return first;
    }
};
